pkg/apis/resource/types.go

```diff
@@ -0,0 +1,449 @@
+	// attempt has definitely failed or b) when the allocated resource was
+	// freed. This ensures that resources are not leaked.
+	metav1.ObjectMeta
+
+	// Spec describes the desired attributes of a resource that then needs
+	// to be allocated. It can only be set once when creating the
+	// ResourceClaim.
+	Spec ResourceClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ResourceClaimStatus
+}
+
+// ResourceClaimSpec defines how a resource is to be allocated.
+type ResourceClaimSpec struct {
+	// ResourceClassName references the driver and additional parameters
+	// via the name of a ResourceClass that was created as part of the
+	// driver deployment.
+	ResourceClassName string
+
+	// ParametersRef references a separate object with arbitrary parameters
+	// that will be used by the driver when allocating a resource for the
+	// claim.
+	//
+	// The object must be in the same namespace as the ResourceClaim.
+	ParametersRef *ResourceClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. "WaitForFirstConsumer" is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. If this is not specified, the resource is
+	// not yet allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. A Pod which references a ResourceClaim which is not
+	// reserved for that Pod will not be started.
+	//
+	// The maximum size is 32.
+	ReservedFor []ResourceClaimUserReference
```

Comment:
```suggestion
	ReservedFor []ResourceClaimConsumerReference
```

“user” has a very common meaning in IT. Now we're looking at documenting this, and I'm seeing the potential impact, I think “consumer” would be better.

Reply:
Okay.